File: bridge/typescript/src/protocol/connection/payload/receiver.ts
import type { BinaryPayload, PayloadChunkNotification } from "../../../generated/protocol/payload.js";
import type { WorkspaceQueryResponse } from "../../../generated/protocol/query.js";
import type { WorkspaceResponse } from "../../../generated/protocol/response.js";

/** Deferred payload chunks received while waiting for a response. */
export class PayloadReceiver {
    readonly #pending = new Map<number, PayloadStream>();
    readonly #completed = new Map<number, Uint8Array>();

    /** Ingest one payload chunk. */
    ingest(chunk: PayloadChunkNotification): void {
        const id = chunk.id[0];
        const total = chunk.total;
        const index = chunk.index;

        if (!Number.isInteger(total) || total <= 0) {
            throw new Error("payload chunk total must be positive");
        }
        if (!Number.isInteger(index) || index < 0 || index >= total) {
            throw new Error("payload chunk index out of range");
        }
        if (chunk.done && index + 1 !== total) {
            throw new Error("payload chunk done marker is inconsistent");
        }

        const stream = this.#pending.get(id) ?? new PayloadStream(total);
        this.#pending.set(id, stream);
        stream.insert(index, chunk.bytes);

        if (stream.isComplete()) {
            this.#pending.delete(id);
            this.#completed.set(id, stream.bytes());
        }
    }

    /** Resolve deferred payloads referenced by one response. */
    resolve(response: WorkspaceResponse): WorkspaceResponse | undefined {
        if (response.kind !== "queryResult") {
            return response;
        }

        const query = this.#resolveQueryResponse(response.query_result);
        if (query === undefined) {
            return undefined;
        }

        return {
            ...response,
            query_result: query,
        };
    }

    #resolveQueryResponse(response: WorkspaceQueryResponse): WorkspaceQueryResponse | undefined {
        if (response.kind === "query") {
            const payload = this.#resolvePayload(response.query.payload);
            if (payload === undefined) {
                return undefined;
            }

            return {
                ...response,
                query: { payload },
            };
        }

        if (response.kind === "queryBatch") {
            const query_batch = [];
            for (const item of response.query_batch) {
                const payload = this.#resolvePayload(item.payload);
                if (payload === undefined) {
                    return undefined;
                }

                query_batch.push({ payload });
            }

            return {
                ...response,
                query_batch,
            };
        }

        return response;
    }

    #resolvePayload(payload: BinaryPayload): BinaryPayload | undefined {
        if (payload.body.kind === "inline") {
            return payload;
        }

        const id = payload.body.id[0];
        const bytes = this.#completed.get(id);
        if (bytes === undefined) {
            return undefined;
        }
        if (bytes.length !== payload.body.totalBytes) {
            throw new Error("payload size mismatch");
        }

        this.#completed.delete(id);

        return {
            body: {
                kind: "inline",
                bytes,
            },
        };
    }
}

/** Pending payload stream. */
class PayloadStream {
    readonly #chunks: Array<Uint8Array | undefined>;
    #received = 0;

    /** Create a payload stream with a fixed chunk count. */
    constructor(total: number) {
        this.#chunks = new Array(total);
    }

    /** Insert one chunk. */
    insert(index: number, bytes: Uint8Array | readonly number[]): void {
        if (this.#chunks[index] !== undefined) {
            throw new Error("duplicate payload chunk");
        }

        this.#chunks[index] = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes);
        this.#received += 1;
    }

    /** Return whether all chunks have arrived. */
    isComplete(): boolean {
        return this.#received === this.#chunks.length;
    }

    /** Merge chunks into one byte buffer. */
    bytes(): Uint8Array {
        let length = 0;
        for (const chunk of this.#chunks) {
            if (chunk === undefined) {
                throw new Error("payload chunks missing at finalize");
            }

            length += chunk.length;
        }

        const bytes = new Uint8Array(length);
        let offset = 0;
        for (const chunk of this.#chunks) {
            if (chunk === undefined) {
                throw new Error("payload chunks missing at finalize");
            }

            bytes.set(chunk, offset);
            offset += chunk.length;
        }

        return bytes;
    }
}

File: bridge/python/src/destack/protocol/connection/payload/receiver.py
from __future__ import annotations

import builtins

from dataclasses import replace

from destack._generated.protocol.envelope import ProtocolNotification
from destack._generated.protocol.notification import WorkspaceNotificationPayloadChunk
from destack._generated.protocol.payload import (
    BinaryPayload,
    PayloadBodyDeferred,
    PayloadBodyInline,
    PayloadChunkNotification,
)
from destack._generated.protocol.query import (
    QueryResponsePayload,
    WorkspaceQueryResponse,
    WorkspaceQueryResponseQuery,
    WorkspaceQueryResponseQueryBatch,
)
from destack._generated.protocol.response import (
    WorkspaceResponse,
    WorkspaceResponseQueryResult,
)
from ..error import ProtocolRequestError


class PayloadReceiver:
    """Deferred payload chunks received while waiting for a response."""

    def __init__(self) -> None:
        self._pending: dict[int, PayloadStream] = {}
        self._completed: dict[int, builtins.bytes] = {}
        self._response: WorkspaceResponse | None = None

    def ingest_notification(self, notification: ProtocolNotification) -> bool:
        """Ingest one notification if it carries payload bytes."""

        payload = notification.payload
        if isinstance(payload, WorkspaceNotificationPayloadChunk):
            self.ingest(payload.payload_chunk)
            return True

        return False

    def ingest(self, chunk: PayloadChunkNotification) -> None:
        """Ingest one payload chunk."""

        if chunk.done and chunk.index + 1 != chunk.total:
            raise ProtocolRequestError("payload chunk done marker is inconsistent")

        stream = self._pending.get(chunk.id)
        if stream is None:
            stream = PayloadStream(chunk.total)
            self._pending[chunk.id] = stream

        stream.insert(chunk.index, bytes(chunk.bytes))
        if stream.is_complete:
            del self._pending[chunk.id]
            self._completed[chunk.id] = stream.bytes()

    def resolve(self, response: WorkspaceResponse) -> WorkspaceResponse | None:
        """Resolve deferred payloads referenced by one response."""

        self._response = response

        return self.resolve_required()

    def resolve_required(self) -> WorkspaceResponse | None:
        """Resolve the stored response if all payload chunks are present."""

        if self._response is None:
            return None

        if not isinstance(self._response, WorkspaceResponseQueryResult):
            return self._response

        query = self._resolve_query_response(self._response.query_result)
        if query is None:
            return None

        return replace(self._response, query_result=query)

    def _resolve_query_response(
        self,
        response: WorkspaceQueryResponse,
    ) -> WorkspaceQueryResponse | None:
        """Resolve one query response payload."""

        if isinstance(response, WorkspaceQueryResponseQuery):
            payload = self._resolve_payload(response.query.payload)
            if payload is None:
                return None

            return replace(response, query=QueryResponsePayload(payload=payload))

        if isinstance(response, WorkspaceQueryResponseQueryBatch):
            batch: list[QueryResponsePayload] = []
            for item in response.query_batch:
                payload = self._resolve_payload(item.payload)
                if payload is None:
                    return None

                batch.append(QueryResponsePayload(payload=payload))

            return replace(response, query_batch=batch)

        return response

    def _resolve_payload(self, payload: BinaryPayload) -> BinaryPayload | None:
        """Resolve one binary payload if its bytes are present."""

        if isinstance(payload.body, PayloadBodyInline):
            return payload

        if not isinstance(payload.body, PayloadBodyDeferred):
            raise ProtocolRequestError("unknown payload body")

        payload_id = payload.body.id
        bytes_ = self._completed.get(payload_id)
        if bytes_ is None:
            return None
        if len(bytes_) != payload.body.total_bytes:
            raise ProtocolRequestError("payload size mismatch")

        del self._completed[payload_id]

        return BinaryPayload(body=PayloadBodyInline(bytes=bytes_))


class PayloadStream:
    """Pending payload stream."""

    def __init__(self, total: int) -> None:
        if not isinstance(total, int) or total <= 0:
            raise ProtocolRequestError("payload chunk total must be positive")

        self._chunks: list[builtins.bytes | None] = [None] * total
        self._received = 0

    @property
    def is_complete(self) -> bool:
        """Return whether all chunks have arrived."""

        return self._received == len(self._chunks)

    def insert(self, index: int, data: builtins.bytes) -> None:
        """Insert one chunk."""

        if not isinstance(index, int) or index < 0 or index >= len(self._chunks):
            raise ProtocolRequestError("payload chunk index out of range")
        if self._chunks[index] is not None:
            raise ProtocolRequestError("duplicate payload chunk")

        self._chunks[index] = data
        self._received += 1

    def bytes(self) -> builtins.bytes:
        """Merge chunks into one byte buffer."""

        chunks = []
        for chunk in self._chunks:
            if chunk is None:
                raise ProtocolRequestError("payload chunks missing at finalize")

            chunks.append(chunk)

        return b"".join(chunks)
